### PR TITLE
nixos/virtualisation/nixos-containers: add systemCallFilter option, remove null checks

### PR DESCRIPTION
**Pull request by Samuel-Martineau**
Add systemCallFilter to NixOS containers. This is notably required to run docker in nspawn. As far as I know, the only current way to do this is [this very ugly hack](https://discourse.nixos.org/t/podman-docker-in-nixos-container-ideally-in-unprivileged-one/22909/12):
```nix
additionalCapabilities = [
  # This is a very ugly hack to add the system-call-filter flag to
  # nspawn. extraFlags is written to an env file as an env var and
  # does not support spaces in arguments, so I take advantage of
  # the additionalCapabilities generation to inject the command
  # line argument.
  ''CAP_NET_ADMIN" --system-call-filter="add_key keyctl bpf" --capability="CAP_NET_ADMIN''
];
```

This PR allows this to be written as
```nix
systemCallFilter = [ "add_key" "keyctl" "bpf" ]
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/virtualisation/nixos-containers.nix [hunk 1]
```diff
@@ -545,6 +548,16 @@ in
               '';
             };
 
+            systemCallFilter = mkOption {
+              type = types.listOf types.str;
+              default = [];
+              example = [ "add_key" "keyctl" "bpf" ];
+              description = lib.mdDoc ''
+                Set the system call filter for the container.  See the
+                systemd-nspawn(1) man page for more information.
```
Comment (h7x4) [inline nixos/modules/virtualisation/nixos-containers.nix:557]:
```suggestion
                {manpage}`systemd-nspawn(1)` man page for more information.
```

File: nixos/modules/virtualisation/nixos-containers.nix [hunk 2]
```diff
@@ -173,6 +173,9 @@ let
         ${optionalString (cfg.additionalCapabilities != null && cfg.additionalCapabilities != [])
           ''--capability="${concatStringsSep "," cfg.additionalCapabilities}"''
         } \
+        ${optionalString (cfg.systemCallFilter != null && cfg.systemCallFilter != [])
```
Comment (h7x4) [inline nixos/modules/virtualisation/nixos-containers.nix:176]: I see this has been done by the other options, but it doesn't make any sense to me. How could `cfg.systemCallFilter` ever be `null` if it has already been typechecked with the type `listOf str` and `[ ]` as default?

Comment (Samuel-Martineau) [reply nixos/modules/virtualisation/nixos-containers.nix:176]: I fully agree with you. I just copied this check from the line just above about `additionalCapabilities`, which has the same type and the same default value. I figured there must be something I didn't understand about validation in Nix. Could you just confirm that this check if unnecessary?

Comment (Samuel-Martineau) [reply nixos/modules/virtualisation/nixos-containers.nix:176]: This check seems to go back to the creation of the `additionalCapabilities` option (0d1e1b1810081f8fb6496e2879bfa8326d79e580). However, even at that time, the option already had the same type and default. This is really a mystery for me.

Comment (h7x4) [reply nixos/modules/virtualisation/nixos-containers.nix:176]: Feel free to just not do it. I really don't believe there's any good reason to. Maybe even make another commit, fixing up the others if you'd like. We have quite a large test suite for these (see `nixos/tests/containers-*.nix`), so it probably shouldn't be a problem if all of them succeed.

**Comment by Samuel-Martineau** [comment]
@h7x4 Please let me know if there is anything else I need to fix.

**Comment by Samuel-Martineau** [comment]
I think everything is working fine now!

**Comment by h7x4** [comment]
Neat! Maybe you could do some git-fu to squash all the refactoring commits into a single commit, and place it before the `add systemCallFilter` commit, so you don't need to introduce the null check only to remove it again?

**Comment by Samuel-Martineau** [comment]
I am not that good with Git, so give me a bit of time to figure out how to do that. This will require a force push to my branch though, right?

**Comment by h7x4** [comment]
Yes, it will require a forcepush as you are now not only appending commits, but rewriting them. Have a look into "interactive rebasing".

Also, just be careful to not rebase 1000 commits (or something) and then forcepush, because everyone will get pinged. This happens regularly, and we have to lock down the PRs and ask people to open new ones because of all the noise it generates. If you mark it as draft first, people won't get pinged until you remark it as ready for release, which gives you a sort of safety net.

**Comment by Samuel-Martineau** [comment]
I think I figured it out. Is this what you had in mind?

**Comment by h7x4** [comment]
Indeed! Cool beans!

**Review by h7x4 — APPROVED** [review]

File: nixos/modules/virtualisation/nixos-containers.nix [hunk 3]
```diff
@@ -545,6 +549,16 @@ in
               '';
             };
 
+            systemCallFilter = mkOption {
+              type = types.listOf types.str;
+              default = [];
+              example = [ "add_key" "keyctl" "bpf" ];
+              description = lib.mdDoc ''
+                Set the system call filter for the container.  See the
```
Comment (h7x4) [inline nixos/modules/virtualisation/nixos-containers.nix:557]: nit:

```suggestion
                Set the system call filter for the container. See the
```

Comment (Samuel-Martineau) [reply nixos/modules/virtualisation/nixos-containers.nix:557]: I just copied the double space from the above:
```nix
additionalCapabilities = mkOption {
  type = types.listOf types.str;
  default = [];
  example = [ "CAP_NET_ADMIN" "CAP_MKNOD" ];
  description = lib.mdDoc ''
    Grant additional capabilities to the container.  See the
    capabilities(7) and systemd-nspawn(1) man pages for more
    information.
  '';
};
```
I hence just assumed Nix manual pages followed [classic English spacing](https://en.wikipedia.org/wiki/History_of_sentence_spacing).

Comment (Samuel-Martineau) [reply nixos/modules/virtualisation/nixos-containers.nix:557]: Let me fix it for both

Comment (h7x4) [reply nixos/modules/virtualisation/nixos-containers.nix]: > classic English spacing

Whoa, today I learned

The nix manual pages doesn't even follow it's own literal spacing. Everything is commonmark and a newline isn't a newline until there's two of them.

**Comment by Samuel-Martineau** [comment]
I think everything is finally alright

**Comment by h7x4** [comment]
Neat!

I don't know how familiar you are with nixpkgs, since this is your first contribution, but as you might see from the other PRs, getting changes merged can sometimes be a bit of a patience game.

There are some threads on discourse where you can ask for more reviews (namely the "PRs ready for *" threads), and there's the different matrix dev channels. You can also try requesting reviews from people you think would like to know about this change (although that might require a bit of familiarity with the project and community). Alternatively, the collectively best option would be to help out by making yourself familiar with the contribution guidelines and try reviewing some PRs yourself.

**Comment by Samuel-Martineau** [comment]
I am not familiar with the process, thanks for enlightening me (and all your help). Patience won't be an issue; as I've said, above, I already have a temporary fix. I'll try to lookup these guidelines

**Review by h7x4 — APPROVED** [review]

**Pull request closed without merging**